Cluster management

broxtronix opened the pull request:
This is a first attempt to improve the `dcluster` command line tool for `distributed`.  This PR includes the following improvements:

 1. `dcluster` now has command line arguments for specifying the number of cores/worker (`--ncores`) and the number of workers/host (`--nworkers`).  This makes it easy to move adjust the blend of SMP vs. multi-process parallelism on hosts in the cluster.  Some workloads benefit from SMP, while others benefit from multi-processing.

 2. The `--center-port` can now be specified in `dcluster`.

 3. The `dworker` port is now selected at random.  However, I see that you just checked in another PR #52  that addresses this issue, so possibly this should be removed here.

 4. A new `--log-directory` argument allows the user to specify a directory to use on each host in the cluster for streaming the `stdout` and `stderr` of all `dcenter` and `dworker` commands to disk.  This has the benefit of creating useful logfiles, and it also reduces the amount of network chatter over the `dcluster` ssh connections, which may be an important optimization for very large clusters with many workers.

 5. If no `--log-directory` is specified (the default), then the output of all `dworker` and `dcenter` processes is streamed to the console running `dworker`.  This output has been cleaned up to make this stream of information more readable and useful.  There is also light use of terminal colors that, if you find it superfluous, would be easy to remove.  (I like the enhanced readability!) :)

  6. The `dworker` command now captures any `KeyboardInterrupt` event, and shuts itself and all remote `dworker` and `dcenter` processes down cleanly.  Note that this means that the cluster lifetime is currently tied to the lifetime of the `dcluster` process.  This is convenient and easy to use, but does mean that `dcluster` must remain running, and any interruptions to ssh connections between the master and worker nodes may cause remote processes to terminate.

Having spent some time today hacking and thinking on this today, I have a few thoughts related to this current tool:

* On a cluster with 32 nodes and 4 hosts, I can potentially run up to 128 `dworker` processes.  Right now, each one requires its own SSH connection, and these take just over a second to establish.  Thus, it can already take a minute to start up a small cluster with many CPUs if I want to use "multiprocessing" mode.  Establishing these connection concurrently (via threading these ssh connections) will probably be necessary to reduce startup time.  For me, this is the next logical addition on top of what I've done here. I'd be happy to work on that if you think it would be a good idea.  

* That said, it would be nice if the `dworker` command could could have a command line argument to switch it from SMP mode to multi-processing mode.  That way, we would still only need one ssh connection per host, and could switch between SMP vs multi-processing workers with a quick command-line switch.

* Although I am actually sort of liking this ability to start and stop a cluster simply by starting and stopping `dcluster`, I think a more robust long-term solution would be to [daemonize](http://www.jejik.com/files/examples/daemon3x.py) the `dcenter` and `dworker` processes so that they can start and run in the background.  This way their lifetime does not depend on the persistence of SSH connections or a single remote command on the master that the user might carelessly terminate.  At that point, the user might start and stop the cluster using `dcluster start` and `dcluster stop`.  

In any case, this is all just spit-balling.  Let me know what you think of the improved `dcluster`, and in particular what you think of these last two larger questions.

broxtronix commented:
Oh, forgot to cc @mrocklin ...

mrocklin commented:
This looks great to me so far.  Hope to play with it in an hour or so.  cc @danielfrg

mrocklin commented:
This looks great.  It's nice to see something more serious come to fruition here.  I'll add some new-user feedback here on speedbumps I ran into.  I'll add some code things inline. 

Running `dcluster` naively results in an uninformative error (this probably is a hold-over error from my initial work before):

```
mrocklin@notebook:~/workspace/distributed$ dcluster 
Traceback (most recent call last):
  File "/home/mrocklin/workspace/distributed/bin/dcluster", line 41, in <module>
    start()
  File "/home/mrocklin/Software/anaconda/lib/python3.4/site-packages/click/core.py", line 664, in __call__
    return self.main(*args, **kwargs)
  File "/home/mrocklin/Software/anaconda/lib/python3.4/site-packages/click/core.py", line 644, in main
    rv = self.invoke(ctx)
  File "/home/mrocklin/Software/anaconda/lib/python3.4/site-packages/click/core.py", line 837, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mrocklin/Software/anaconda/lib/python3.4/site-packages/click/core.py", line 464, in invoke
    return callback(*args, **kwargs)
  File "/home/mrocklin/workspace/distributed/bin/dcluster", line 28, in start
    center, hostnames = hostnames[0], hostnames[1:]
IndexError: list index out of range
```

On my machine I don't immediately find the `distributed` package.  I think that this is because we haven't invoked my bashrc file.  I think that my solution to this previously was to use `invoke_shell` but I could be wrong.  Presumably this isn't so important for the actual cluster case, but is nice to support for local testing.

```
[ center 127.0.0.1:8787 ] : remote process exited with exit status 1
[ center 127.0.0.1:8787 ] : Traceback (most recent call last):
[ center 127.0.0.1:8787 ] :   File "/usr/local/bin/dcenter", line 4, in <module>
[ center 127.0.0.1:8787 ] :     import pkg_resources
[ center 127.0.0.1:8787 ] :   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 2749, in <module>
[ center 127.0.0.1:8787 ] :     working_set = WorkingSet._build_master()
[ center 127.0.0.1:8787 ] :   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 444, in _build_master
[ center 127.0.0.1:8787 ] :     ws.require(__requires__)
[ center 127.0.0.1:8787 ] :   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 725, in require
[ center 127.0.0.1:8787 ] :     needed = self.resolve(parse_requirements(requirements))
[ center 127.0.0.1:8787 ] :   File "/usr/lib/python3/dist-packages/pkg_resources.py", line 628, in resolve
[ center 127.0.0.1:8787 ] :     raise DistributionNotFound(req)
[ center 127.0.0.1:8787 ] : pkg_resources.DistributionNotFound: distributed==0.0.1
```

mrocklin commented on distributed/cluster.py:
```diff
@@ -2,65 +2,196 @@
+    if logdir is not None:
+        cmd = 'mkdir -p {logdir} && ' + cmd
+        cmd += '&> {logdir}/dcenter_{addr}:{port}.log\n'.format(
+            addr = center_addr, port = center_port, logdir = logdir)
```
I'm not sure if it makes sense to use the logging module directly to direct logs to a file.  Presumably this would mean adding extra logic into `dcenter/dworker` to change logging levels and add `logging.FileHandlers`.  @quasiben you've played a lot with logging recently.  Is there is any benefit one way or the other?

mrocklin commented on distributed/cluster.py:
```diff
@@ -2,65 +2,196 @@
+        cmd = 'dworker {center_addr}:{center_port} --host {worker_addr} --port {worker_port} --ncores {ncpu}\n'.format(
+            center_addr = center_addr, center_port = center_port,
+            worker_addr = worker_addr, worker_port = worker_port,
+            ncpu = cpus_per_worker, logdir = logdir)
```
Perhaps avoid duplicated code by just adding the `--ncores` keyword in a smaller if branch

mrocklin commented:
Concurrent ssh connections would certainly be pleasant.  It's unfortunate that there isn't a simple SSH library on top of Tornado.  This is actually the first time that using `asyncio` would have been more appropriate http://asyncssh.readthedocs.org/en/latest/ .  That being said, threads are probably completely satisfactory here.

broxtronix commented:
Ok, I addressed most of your comments, I think.   Important changes:

* I switched back to using `invoke_shell()`, since this seems to more reliably invoke an interactive shell that loads all of the users normal shell dot files.  I would have thought that `exec_command()` with `use_pty=True` would have done this as well, but it seems like that isn't the case.  Invoking a full shell is probably going to be the most robust to varied user environments, so it probably makes sense just to go with that.

* I improved the help text when you run `dcluster` with no arguments.

* I made a slight tweak to the syntax of the command line.  Now if you specify multiple hosts, and don't specify `--center` explicitly, then both a `dcenter` and `dworker` are run on the first host in the list.  (Before the `dcenter` was run on the first host, and `dworker`s on all the rest, which I found confusing and somewhat redundant with the `--center` option.

* Other cleanup to address your code comments.

* Addressing your comment about logging above, I think it would generally be better to use a dedicated logger in the code itself, rather than redirecting output, but that path involves some more work.  Curious what @quasiben has to say.

Regarding threads, I'm tempted to take a quick crack at that right now.  (Alas, asyncssh looks nice, but threads will work great for this too.)

broxtronix commented:
For starting clusters, I wanted to point out two projects that may provide some inspiration / food for thought.  My [spark-gce](https://github.com/broxtronix/spark-gce) project starts a Spark cluster on Google Compute Engine, and [Flintrock](https://github.com/nchammas/flintrock) performs a similar function on Amazon EC2.  They are quite similar at their core... by and large they are scripts for managing instances and their resources.  They install a few software packages to set up a sane development environment, but this can be easily customized.  They would translate over with very little modification to starting a distributed cluster.

broxtronix commented:
Hmm, before I dive into multi-threading the SSH connections, I'm curious to ask what method of Python parallelism you would use for this.  Would a `multiprocessing.Pool` be good with `multiprocessing.manager.Queue()` objects for interprocess communication?  You're the expert, so I'm curious what you think is best!

mrocklin commented:
I would probably use threads.  I wouldn't use a pool.  Something like the following?

```python
from queue import Queue
from threading import Thread
d = dict()
for worker in workers:
    queue = Queue()
    thread = Thread(target=start_worker, args=(queue, ...))
    thread.start()
    d[worker] = {'queue': queue, 'thread': thread}
```

mrocklin commented:
Basic local use is pretty slick.  Way nicer than before.

Some additional usability comments.

1.  We could consider extending `dcluster` to include a `dscheduler` process on the same node provided for the center
2.  It would be nice to have a highly visible line that gives the address of the center or scheduler.  This would need to be visibly marked so that it didn't fall into the mess of logs

    ```
    ==============
    Center Address:
    Scheduler Address: 
    ==============
    ```

    I often search for such an address and then copy-paste it into a call to `Executor`
3.  How should we specify options to SSH?  For example I often use a special ssh key when interacting with nodes on particular clusters and use a special username.  I might do something like `ssh -i ~/.ssh/my-special-key.pem ubuntu@hostname`.

danielfrg commented:
This is really nice work!

I was planning on creating an easy `dec2` script to create instances in ec2 and install the requirements for `distributed` to work but if @broxtronix is planning on doing this also i might just let you do it :)

My plan was to based it on apache-libcloud so it can be easily extended to more clouds than just AWS and also based it on [salt](https://github.com/danielfrg/datasciencebox) more specifically salt-ssh so people don't need to install salt in their nodes they just need the python salt-ssh package in their laptop.

Thats the approach that I have used to create my mesos/spark launcher: https://github.com/danielfrg/datasciencebox but it might be to much for distributed since its a simpler deployment.

mrocklin commented:
@danielfrg does salt still have a Python 3 issue?  Any particular thoughts on spark-gce or flintrock?

danielfrg commented:
Yes, `salt-ssh` do require python 2.

Both spark-gce and flintrock look really good, the only thing I would change is boto for apache-libcloud to make it easier to deploy in multiple clouds (eventually, ec2 for sure should be the first target) for example: https://github.com/ema/nubo is a cli for libcloud.

broxtronix commented:
Alrighty, here is a fully threaded version.  Startup is definitely a lot faster, and everything else works more or less as before.  I did switch back to `exec_command()` because this seems to be the only way to get proper stderr streams and exit status reporting working.  However, I think I found a way to launch commands through an (interactive login)[http://capistranorb.com/documentation/faq/why-does-something-work-in-my-ssh-session-but-not-in-capistrano/#] shell, which seems to pull in the necessary environment variables.

I also added the startup text you suggested.  Good idea!

Anyway, take this out for a spin and let me know what you think.

I need to run to dinner, but I'll do a bit of research on `salt-ssh` and `apache-libcloud` to familiarize myself with them, as I haven't used them before.  More soon.

mrocklin commented on distributed/cluster.py:
```diff
@@ -2,65 +2,252 @@
+            while(1):
+                for process in all_processes:
+                    while not process['output_queue'].empty():
+                        print(process['output_queue'].get())
```
I was thinking about this last night.  It might be better to have a single output_queue that all threads write to.  They're threadsafe and so you can just do 

```python
while true:
    print(queue.get())
```

mrocklin commented on distributed/cluster.py:
```diff
@@ -2,65 +2,252 @@
+
+                # Kill some time and free up CPU before starting the next sweep
+                # through the processes.
+                sleep(1.0)
```
Can we reduce this to `100ms` without significant cost?

mrocklin commented:
This works quite well for me interactively.

broxtronix commented:
Reducing the sleep times a bit did not seem to incur much additional CPU overhead, so I want ahead and checked in that change.  For one vs. many output queues, I am definitely glad to make that change as well, but before I do I wanted to raise two possible concerns I had with doing it that way.  I'm curious to hear your feedback, and can then execute whichever decision you reach.

My primary worry is that having hundreds (or potentially even thousands) of threads posting output to a single queue might create some resource contention as they wait for each other to lock and unlock the queue.  There is a lot of output generated when things really get going, and I can easily imagine that a single queue would be easily overwhelmed, and the `dcluster` node may not even be able to drain it fast enough with so many processes feeding it fresh lines to display.  

The other benefit to having separate queues is that you can drain them one at a time, which helps to wrangle the lines of output from one remote worker process roughly adjacent to each other in the `dcluster` output, rather than being heavily interleaved.  

As I raise these concerns, I'm realizing that there may be other ways they could be addressed.  For example, I could aggregate multiple lines of output before posting to the queue in `async_ssh()`, thereby reducing the number of times the queue needs to be locked/unlocked and helping to aggregate lines of output.  That's probably a good thing to do in any case. 

The other thing that would help here would be if `dworker` had a `--multiprocessing` argument (or similar) that caused it to launch multiple workers as separate processes.  This would reduce the number of SSH connections needed when I am electing to using multiprocess parallelism rather than threading parallelism on the worker nodes.  Fewer SSH connections would mean fewer queues as well.  Is a `multiprocessing` mode something that `dworker` could support, or is it easier to simply start many `dworker` processes over SSH?

mrocklin commented:
The reasons behind multiple queues and second long delays make sense.  A one second delay also doesn't look bad in this context. 

Also, I'm definitely not executive here; you're way more qualified here than I am.  I'm really hoping that you'd like to take on the role of deployment czar.

Adding `--multiprocessing` to `dworker` would make this (and other things) easier.  It'll add a bit of pain to `dworker` but it's probably worth it.  I'll take a look at this this afternoon.

broxtronix commented:
I think that the 0.1 second delay on the consumer of the queue and 1 second delays on the producers should help to keep the data flowing through the pipes without any blockages.  If you are ok with the multiple queues, then let's stick with that for now.  Thanks for taking the time to digest my response, and for looking further into the `--multiprocessing` flag for `dworker`.

I will spend some time today thinking about deployment and reading through the links that @danielfrg posted.  I must say, @danielfrg, your [datasciencebox](https://github.com/danielfrg/datasciencebox) project looks very elegant.  I'm eager to learn more about it!  More on all this soon.

broxtronix commented:
Ok, this latest checkin has transitioned `dcluster` over to using the new `dworker` command line API from #57 .  It starts just one `dworker` process on each host, and uses the `--nthreads` and `--nprocs` arguments to control threading vs multiprocessing parallelism.  

However, when testing it with `--nprocs` larger than 1, it did not seem to work as I had expected.  It seems to only start one process per `dworker`.   For example:
```
dcluster 127.0.0.1 --nthreads 1 --nprocs 4
```
runs a single `dworker` like this
```
dworker 127.0.0.1:8787 --host 127.0.0.1 --port 17516 --nthreads 1 --nprocs 4
```
The connected executor shows only one core and one process.
```
>>> from distributed import Executor
>>> executor = Executor('127.0.0.1:8787')
>>> executor.scheduler.ncores
{('127.0.0.1', 17516): 1}
```

broxtronix commented:
Perhaps this is just a mistake in how I'm interpreting the output of `executor.scheduler.ncores`?  I'm looking at your code in `dworker` and everything seems copacetic.

mrocklin commented:
OK, the problem here is in specifying the port explicitly.  `dworker` calls `Nanny` many times, each with the port argument.  In all of my testing I always played with `port=0` and so the workers chose random ports.  When we specify a single port it's not clear what we should do.

Probably `dworker` should err if `nprocs > 1 and port != 0`.  Probably `dcluster` should not specify ports on individual workers.

broxtronix commented:
Ah, that makes sense.  I made those changes (including the error checking in `dworker`) and all seems to be working great again now.

mrocklin commented:
Happy to merge this any time if you'd like to sync up with master.

The next things I would personally need to use this with my existing clusters is to forward arguments along to `paramiko.SSHClient.connect`, in particular `username`, `pkey`, and sometimes even `port` end up being useful to me.

broxtronix commented:
Oh, right!  Adding the SSH options will be straight-forward.  I can do that first thing tomorrow, and then it would be awesome to merge this into master.  Thanks for all the constructive advice and help on this PR.  `distributed` has been working great for me over the past couple of days – I'm using it on a 4 host x 32 core cluster, and everything is humming along great.

broxtronix commented:
Alrighty, the SSH options are new broken out in the form of these arguments:
* `--ssh-username`
* `-ssh-port`
* `-ssh-private-key`

These do what you would expect.  I tested it a bit on my systems, but I'm curious to hear if it works on your clusters.

I also improved the SSH error handling code to provide slightly more insightful error messages when one of these parameters has a bogus value.  And now, if one of the commands fails after three retries, `dcluster` will exit to the shell.  I also turned off `INFO` logging in Paramiko, which seemed a little too verbose to be useful.  (We can turn that back on if you prefer more verbose output.)  

Just a quick note: supplying an incorrect port number will currently trigger a Paramiko exception that is not very informative that is described here: https://github.com/paramiko/paramiko/issues/615.  However, seems like that bug is fixed and will roll out in future versions of their library.

mrocklin commented:
Cool.  I'll give this a quick spin tonight and probably merge then.  This is great!

mrocklin commented:
Merged.  Looking forward to using this.  Thanks @broxtronix !
